File: app/src/test/java/net/reichholf/dreamdroid/ui/current/CurrentServiceMapperTest.java
package net.reichholf.dreamdroid.ui.current;

import net.reichholf.dreamdroid.enigma.CurrentService;
import net.reichholf.dreamdroid.enigma.CurrentServiceParser;
import net.reichholf.dreamdroid.enigma.Event;
import net.reichholf.dreamdroid.helpers.ExtendedHashMap;
import net.reichholf.dreamdroid.ui.epg.EpgListMapper;

import org.junit.Test;

import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;

public class CurrentServiceMapperTest {
    @Test
    public void eventToExtendedHashMapCopiesEventFields() throws Exception {
        InputStream in = getClass().getResourceAsStream("/web/getcurrent.xml");
        assertNotNull(in);
        String xml = new String(in.readAllBytes(), StandardCharsets.UTF_8);
        CurrentService current = CurrentServiceParser.INSTANCE.parse(xml);
        assertNotNull(current);
        Event event = current.getNow();
        assertNotNull(event);

        ExtendedHashMap map = CurrentServiceMapper.eventToExtendedHashMap(event);
        assertNotNull(map);
        assertEquals(event.getEventId(), map.getString(net.reichholf.dreamdroid.helpers.enigma2.Event.KEY_EVENT_ID));
        assertEquals(event.getTitle(), map.getString(net.reichholf.dreamdroid.helpers.enigma2.Event.KEY_EVENT_TITLE));
        assertEquals(EpgListMapper.toExtendedHashMap(event).getString(
                net.reichholf.dreamdroid.helpers.enigma2.Event.KEY_EVENT_START_READABLE),
                map.getString(net.reichholf.dreamdroid.helpers.enigma2.Event.KEY_EVENT_START_READABLE));
    }

    @Test
    public void eventToExtendedHashMapNullIsNull() {
        assertNull(CurrentServiceMapper.eventToExtendedHashMap(null));
    }

    @Test
    public void toExtendedHashMapCopiesServiceAndEvents() throws Exception {
        InputStream in = getClass().getResourceAsStream("/web/getcurrent.xml");
        assertNotNull(in);
        String xml = new String(in.readAllBytes(), StandardCharsets.UTF_8);
        CurrentService current = CurrentServiceParser.INSTANCE.parse(xml);
        assertNotNull(current);

        ExtendedHashMap map = CurrentServiceMapper.toExtendedHashMap(current);
        ExtendedHashMap service = (ExtendedHashMap) map.get(
                net.reichholf.dreamdroid.helpers.enigma2.CurrentService.KEY_SERVICE);
        assertNotNull(service);
        assertEquals(current.getService().getReference(),
                service.getString(net.reichholf.dreamdroid.helpers.enigma2.Service.KEY_REFERENCE));
        assertEquals(current.getService().getProvider(),
                service.getString(net.reichholf.dreamdroid.helpers.enigma2.CurrentService.KEY_SERVICE_PROVIDER));

        @SuppressWarnings("unchecked")
        ArrayList<ExtendedHashMap> events = (ArrayList<ExtendedHashMap>) map.get(
                net.reichholf.dreamdroid.helpers.enigma2.CurrentService.KEY_EVENTS);
        assertNotNull(events);
        assertEquals(2, events.size());
        assertEquals(current.getNow().getEventId(),
                events.get(0).getString(net.reichholf.dreamdroid.helpers.enigma2.Event.KEY_EVENT_ID));
        assertEquals(current.getNext().getEventId(),
                events.get(1).getString(net.reichholf.dreamdroid.helpers.enigma2.Event.KEY_EVENT_ID));
    }
}

File: app/androidTest/java/net/reichholf/dreamdroid/ui/current/CurrentServiceMapperTest.kt
package net.reichholf.dreamdroid.ui.current

import androidx.test.ext.junit.runners.AndroidJUnit4
import net.reichholf.dreamdroid.enigma.CurrentServiceParser
import net.reichholf.dreamdroid.helpers.ExtendedHashMap
import net.reichholf.dreamdroid.testutil.loadWebFixture
import net.reichholf.dreamdroid.ui.epg.EpgListMapper
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class CurrentServiceMapperTest {
    @Test
    fun eventToExtendedHashMapCopiesEventFields() {
        val current = CurrentServiceParser.parse(loadWebFixture("getcurrent.xml"))
        assertNotNull(current)
        val event = current!!.now
        assertNotNull(event)

        val map = CurrentServiceMapper.eventToExtendedHashMap(event)
        assertNotNull(map)
        assertEquals(event!!.eventId, map!!.getString(net.reichholf.dreamdroid.helpers.enigma2.Event.KEY_EVENT_ID))
        assertEquals(event.title, map.getString(net.reichholf.dreamdroid.helpers.enigma2.Event.KEY_EVENT_TITLE))
        assertEquals(
            EpgListMapper.toExtendedHashMap(event).getString(
                net.reichholf.dreamdroid.helpers.enigma2.Event.KEY_EVENT_START_READABLE,
            ),
            map.getString(net.reichholf.dreamdroid.helpers.enigma2.Event.KEY_EVENT_START_READABLE),
        )
    }

    @Test
    fun eventToExtendedHashMapNullIsNull() {
        assertNull(CurrentServiceMapper.eventToExtendedHashMap(null))
    }

    @Test
    fun toExtendedHashMapCopiesServiceAndEvents() {
        val current = CurrentServiceParser.parse(loadWebFixture("getcurrent.xml"))
        assertNotNull(current)

        val map = CurrentServiceMapper.toExtendedHashMap(current)
        val service = map[net.reichholf.dreamdroid.helpers.enigma2.CurrentService.KEY_SERVICE] as ExtendedHashMap
        assertNotNull(service)
        assertEquals(
            current!!.service.reference,
            service.getString(net.reichholf.dreamdroid.helpers.enigma2.Service.KEY_REFERENCE),
        )
        assertEquals(
            current.service.provider,
            service.getString(net.reichholf.dreamdroid.helpers.enigma2.CurrentService.KEY_SERVICE_PROVIDER),
        )

        @Suppress("UNCHECKED_CAST")
        val events = map[net.reichholf.dreamdroid.helpers.enigma2.CurrentService.KEY_EVENTS]
            as ArrayList<ExtendedHashMap>
        assertNotNull(events)
        assertEquals(2, events.size)
        assertEquals(
            current.now!!.eventId,
            events[0].getString(net.reichholf.dreamdroid.helpers.enigma2.Event.KEY_EVENT_ID),
        )
        assertEquals(
            current.next!!.eventId,
            events[1].getString(net.reichholf.dreamdroid.helpers.enigma2.Event.KEY_EVENT_ID),
        )
    }
}
